File: javascript/kth-element.spec.js
'use strict';

var expect = require('chai').expect;
var jsc = require("jsverify");
var KthElement = require('./kth-element.js');

describe('Kth Element in Linear Time', function() {
  it('should exist', function() {
    expect(KthElement).to.not.be.undefined;
  });
  it('provides a working O(n log n) implementation using sort', function() {
    expect(KthElement.cheat([2,4,1,0,9], 1)).to.equal(0);
  });
  it('provides a working O(n) implementation without sort', function() {
    expect(KthElement.find([2,4,1,0,9], 1)).to.equal(0);
  });
  describe("find has cheat's correctness", function () {
    jsc.property("idempotent", "array nat", function (arr) {
      return KthElement.find(arr, 2) === KthElement.cheat(arr, 2);
    });
  });
});

File: ruby/kth_element_spec.rb
require 'rantly/rspec_extensions'
require './kth_element'

describe KthElement do
  it 'gives Array a .find method' do
    expect(KthElement.respond_to?(:find)).to be true
  end
  it 'gives Array a .cheat method' do
    expect(KthElement.respond_to?(:cheat)).to be true
  end

  describe '.cheat' do
    it 'returns kth element statistic in O(n log n)' do
      expect(KthElement.cheat([2,1,3], 2)).to eq(2)
    end
  end

  describe '.find' do
    it 'returns kth element statistic in O(n)' do
      expect(KthElement.find([2,1,3], 2)).to eq(2)
    end
    it 'returns same answers as .cheat for same inputs' do
      property_of {
        len = Rantly { call([:range,1,100]) }
        array(len){integer}
      }.check { |arr|
        expect(KthElement.find(arr, 2)).to eq(KthElement.cheat(arr, 2))
      }
    end
  end
end
